test/jdk/jdk/nio/zipfs/ZipFSDirectoryExceptionMessageTest.java

```diff
@@ -0,0 +1,99 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import java.io.IOException;
+import java.nio.file.*;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipOutputStream;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+/**
+ * @test
+ * @bug 8305945
+ * @summary Validate that Zip FS provides the correct exception message when an
+ * attempt is made to obtain an InputStream from a directory entry
+ * @modules jdk.zipfs
+ * @run junit ZipFSDirectoryExceptionMessageTest
+ */
+public class ZipFSDirectoryExceptionMessageTest {
+    /**
+     * Name of Directory created within the Zip file
+     */
+    public static final String DIRECTORY_NAME = "folder/";
+    /**
+     * The expected error message
+     */
+    public static final String DIR_EXCEPTION_MESSAGE = "/folder: is a directory";
+    /**
+     * Zip file to create
+     */
+    public static final Path ZIP_FILE = Path.of("directoryExceptionTest.zip");
+
+    /**
+     * Create a Zip file which contains a single directory entry
+     * @throws IOException if an error occurs
+     */
+    @BeforeEach
+    public void setup() throws IOException {
+        var ze = new ZipEntry(DIRECTORY_NAME);
+        ze.setMethod(ZipEntry.STORED);
+        ze.setCompressedSize(0);
+        ze.setSize(0);
+        ze.setCrc(0);
+        try (ZipOutputStream zos = new ZipOutputStream(
+                Files.newOutputStream(ZIP_FILE))) {
+            zos.putNextEntry(ze);
+        }
+    }
+
+    /**
+     * Delete the Zip file used by the test
+     * @throws IOException If an error occurs
+     */
+    @AfterEach
+    public void cleanup() throws IOException {
+        Files.deleteIfExists(ZIP_FILE);
+    }
+
+    /**
+     * Validate that Zip FS returns the correct Exception message when
+     * attempting to obtain an InputStream from a path representing a directory
+     *
+     * @throws IOException If an error occurs
+     */
+    @Test
+    public void testException() throws IOException {
+        try (FileSystem zipfs = FileSystems.newFileSystem(ZIP_FILE)) {
+            var file = zipfs.getPath(DIRECTORY_NAME);
+            var x = assertThrows(FileSystemException.class, () -> Files.newInputStream(file));
+            assertEquals(DIR_EXCEPTION_MESSAGE, x.getMessage());
```

Comment:
FYI, test doesn't need to depend the exception message. Checking FileSystemException::getOtherFile returns null would be a more robust way of checking that it didn't provide a second file by mistake, e.g.
```
       try {
            Files.newInputStream(DIRECTORY_NAME);
            fail();
        } catch (FileSystemException e) {
            assertNull(e.getOtherFile());
        } catch (IOException ioe) { 
            // allowed
        }
```

Reply:
added validation for other file being null but also kept the existing validation of the message

Reply:
Okay, although I assume this test will fail if it throws a more general IOException (it's allowed to do that) or there is any adjustment to the error message.

Reply:
Yes, in the unlikely event that either  FileSystemException or  the code below from zipfs is changed, then yes the error message could evolve

```

            if (e.isDir())
                throw new FileSystemException(getString(path), null, "is a directory"); 
```

I do not have a strong preference once way or the other and am happy to remove the message validation if that is what you prefer.  I kept the validation as the original issue was surrounding the actual message being generated which aligns with the text included by zipfs when it threw the exception when obtaining the inputstream

I think it is worth mentioning that we have other tests that also validate exception messages(and we recently updated a couple zip tests because the message changed).    So if we are going to move in the direction of not validating exception messages in tests, we should probably consider adding verbiage to the developers guide to provide that guidance and be consistent within the test suite

Let me know your preference and thank you for your thoughts as it is a useful discussion

Reply:
> What you have is fine, I'm just making the point that the exception message is somewhat secondary

I guess this is also a question of testing the interface vs. testing the implementation.

To get good coverage of validation scenarios, I often find it quite useful to test on the actual message. This makes the test overspecific to the interface, but it helps make sure that each error condition is covered independently. Without asserting on the actual message, it can be hard to know if you are being shadowed by another higher level error condition. I've seen this a lot when adding test coverage for various ZIP processing code.